app/services/control_rate_service.rb

```diff
@@ -0,0 +1,93 @@
+class ControlRateService
+  PERCENTAGE_PRECISION = 1
+
+  # Can be initialized with _either_ a date range or a single date to calculate
+  # control rates. Note that for the date range the returned values will be for each month going back
+  # to the beginning of registrations for the region.
+  def initialize(region, range: nil, date: nil)
+    raise ArgumentError, "Cannot provide both a range and date" if range && date
+    raise ArgumentError, "Must provide either a range or a single date" if range.nil? && date.nil?
+    @region = region
+    @facilities = if @region.respond_to?(:facilities)
+      @region.facilities
+    else
+      [region]
+    end
+    @range = range
+    @date = date
+    @end_of_date_range = date || range.end
+    logger.info "#{self.class} created for range: #{range} region: #{region.id} #{region.name}"
+  end
+
+  delegate :logger, to: Rails
+  attr_reader :date
+  attr_reader :facilities
+  attr_reader :range
+  attr_reader :region
+
+  def call
+    data = {
+      controlled_patients: {},
+      controlled_patients_rate: {},
+      registrations: {}
+    }
+
+    data[:cumulative_registrations] = registrations(@end_of_date_range)
+    registration_counts.each do |(date, count)|
+      formatted_period = date.to_s(:month_year)
+      data[:controlled_patients][formatted_period] = controlled_patients(date).count
+      data[:controlled_patients_rate][formatted_period] = percentage(controlled_patients(date).count, count)
+      data[:registrations][formatted_period] = count
+    end
+    data
+  end
+
+  def registrations(time)
+    registration_counts[time.beginning_of_month.to_date]
+  end
+
+  # Calculate all registration counts for entire range, or for the single date provided
+  def registration_counts
+    @registration_counts ||= if range
+      region.patients.with_hypertension
```

Comment:
A region's expected to respond to `patients`, I assume? What kinds of regions are we going to support? For example, if we're using a facility or facility group (district), could we use a more explicit association? Eg. `region.registered_patients`? 

Hmm, I see that the `patients` association on facility groups does pull registered patients. I know it sounds silly, but could we set up another duplicate association called `registered_patients` that does the same thing? It gives us two benefits:
* Consistent and specific interface for this service to compute registrations
* Sets us up nicely to leave this service alone when we switch other touchpoints on the dashboard to start using `assigned_facility` for things.

Reply:
Notably, the `Facility` model has both `patients` and `registered_patients` associations, and I think we want to use the latter here to compute control rate. The former returns all patients that have ever visited the facility. In fact, could you point me to the precise definition for this control rate? I want to make sure we're on the same page.

Reply:
@harimohanraj89 you can find the definitions at this doc: https://docs.google.com/document/d/1MFRYAuy7pgXFZSH_p0YMLAIyGUldXB0KFYZg2jt8KiY/edit#heading=h.m8132ooz7gej -- look under the "Reports: State Level" heading for these new report definitions.

Reply:
One thing to note is that we will want to get the patient status into the views as well, I'll add that as a card to update the materialized view(s) i'm using.  Here is that new story: https://app.clubhouse.io/simpledotorg/story/282/add-necessary-fields-to-materialized-views-for-district-reports